File: apps/wearos/app/src/main/java/com/togplay/vaimarcia/wear/playback/RelayPlaybackDispatcher.kt
package com.togplay.vaimarcia.wear.playback

import com.togplay.vaimarcia.wear.data.companion.WearCompanionClient
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Cenário B: envia o comando de reprodução para o celular pela conexão persistente do Data
 * Layer. O celular já tem o áudio em cache e o reproduz na caixa Bluetooth pareada — da
 * perspectiva do usuário, isso é indistinguível do Cenário A.
 */
@Singleton
class RelayPlaybackDispatcher @Inject constructor(
    private val companionClient: WearCompanionClient,
) : PlaybackEngine {

    override suspend fun play(audioId: String): PlaybackResult {
        val delivered = companionClient.sendPlayCommand(audioId)
        return if (delivered) {
            PlaybackResult.Started
        } else {
            PlaybackResult.Failed("phone_unreachable")
        }
    }
}

File: apps/watchos/VaiMarciaWatch Watch App/Playback/RelayPlaybackDispatcher.swift
import Foundation

/// Cenário B: envia o comando de reprodução para o celular pela sessão persistente do
/// `WatchConnectivity`. O celular já tem o áudio em cache e o reproduz na caixa Bluetooth
/// pareada — da perspectiva do usuário, isso é indistinguível do Cenário A.
final class RelayPlaybackDispatcher: PlaybackEngine {
    private let connectivity: WatchConnectivityManager

    init(connectivity: WatchConnectivityManager = .shared) {
        self.connectivity = connectivity
    }

    func play(audioId: String, completion: @escaping (PlaybackResult) -> Void) {
        connectivity.sendPlayCommand(audioId: audioId) { delivered in
            completion(delivered ? .started : .failed(reason: "phone_unreachable"))
        }
    }
}
